src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs

```diff
@@ -0,0 +1,282 @@
+    {
+        return _semaphore.ExecuteAsync(DisposeCoreInternalAsync);
+
+        Task DisposeCoreInternalAsync()
+        {
+            foreach (HotReloadSessionState sessionState in _activeSessionStates)
+            {
+                sessionState.Process?.Dispose();
+            }
+
+            _activeSessionStates.Clear();
+
+            return Task.CompletedTask;
+        }
+    }
+
+    private void WriteOutputMessage(HotReloadLogMessage hotReloadLogMessage, CancellationToken cancellationToken) => _hotReloadDiagnosticOutputService.Value.WriteLine(hotReloadLogMessage, cancellationToken);
+
+    /// <summary>
+    /// Checks if the project meets the basic requirements to support Hot Reload. Note that there may be other, specific
+    /// settings that prevent the use of Hot Reload even if the basic requirements are met.
+    /// </summary>
+    private async ValueTask<bool> ProjectSupportsHotReloadAsync()
+        => _configuredProject.Capabilities.AppliesTo("SupportsHotReload") &&
+           await _configuredProject.GetProjectPropertyBoolAsync(ConfiguredBrowseObject.DebugSymbolsProperty) &&
+           !await _configuredProject.GetProjectPropertyBoolAsync(ConfigurationGeneral.OptimizeProperty) &&
+           await _configuredProject.GetProjectPropertyValueAsync(ConfigurationGeneral.TargetFrameworkProperty) is not "";
+
+    /// <summary>
+    /// Returns whether or not the project supports startup hooks. These are used to start Hot Reload in the launched process.
+    /// </summary>
+    private async ValueTask<bool> ProjectSupportsStartupHooksAsync()
+        => await _configuredProject.GetProjectPropertyBoolAsync(ConfigurationGeneral.StartupHookSupportProperty, defaultValue: true);
+
+    /// <inheritdoc />
+    public Task ApplyHotReloadUpdateAsync(Func<IDeltaApplier, CancellationToken, Task> applyFunction, CancellationToken cancellationToken)
+    {
+        return _semaphore.ExecuteAsync(ApplyHotReloadUpdateInternalAsync, cancellationToken);
+
+        async Task ApplyHotReloadUpdateInternalAsync()
+        {
+            // Run the updates in parallel
+            List<Task>? updateTasks = null;
+
+            foreach (HotReloadSessionState sessionState in _activeSessionStates)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (sessionState.Session is IProjectHotReloadSessionInternal { DeltaApplier: {} deltaApplier })
+                {
+                    updateTasks ??= [];
+                    updateTasks.Add(applyFunction(deltaApplier, cancellationToken));
+                }
+            }
+
+            // Wait for their completion
+            if (updateTasks is not null)
+            {
+                await Task.WhenAll(updateTasks);
+            }
+        }
+    }
+
+    public Task ActivateSessionAsync(IVsLaunchedProcess launchedProcess, VsDebugTargetProcessInfo vsDebugTargetProcessInfo)
+    {
+        Assumes.NotNull(_pendingSessionState);
```

Comment:
Worth including a message here to make triaging any violations of this easier.


```suggestion
        Assumes.NotNull(_pendingSessionState, "No pending hot reload session to activate.");
```